utils/cautils/tpm.go

```diff
@@ -0,0 +1,651 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed getting EK public key ID: %w", err)
+	}
+	ekKeyURL := ekURL(ekKeyID)
+
+	ekHexFingerprint, err := keyutil.EncodedFingerprint(ekPublic, keyutil.HexFingerprint)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating EK fingerprint: %w", err)
+	}
+
+	// strip off "<hash>:"
+	ekHexFingerprint = strings.Split(ekHexFingerprint, ":")[1]
+
+	// look for an AK named after the EK hex fingerprint by default
+	var ak *tpm.AK
+	if ak, err = t.GetAK(ctx, ekHexFingerprint); err != nil {
+		// return early if an error occurred that doesn't indicate that the AK does not exist
+		if !errors.Is(err, tpm.ErrNotFound) {
+			return nil, fmt.Errorf("failed getting AK: %w", err)
+		}
+		// create a new AK if it wasn't found. We're using the identifier as the name
+		// used for storing the AK for convenience.
+		if ak, err = t.CreateAK(ctx, ekHexFingerprint); err != nil {
+			return nil, fmt.Errorf("failed creating AK: %w", err)
+		}
+	}
+
+	// check if a (valid) AK certificate (chain) is available. Perform attestation flow otherwise.
+	akChain := ak.CertificateChain()
+	if len(akChain) == 0 || !hasValidIdentity(ak, ekKeyURL) {
+		if akChain, err = ac.performAttestation(ctx, t, ak); err != nil {
+			return nil, fmt.Errorf("failed performing AK attestation: %w", err)
+		}
+		if err := ak.SetCertificateChain(ctx, akChain); err != nil {
+			return nil, fmt.Errorf("failed storing AK certificate chain: %w", err)
+		}
+	}
+
+	// when a new certificate was issued for the AK, it is possible the
+	// certificate that was issued doesn't include the expected and/or required
+	// identity, so this is checked before continuing.
+	if !hasValidIdentity(ak, ekKeyURL) {
+		return nil, fmt.Errorf("AK certificate (chain) not valid for EK %q", ekKeyURL)
+	}
+
+	// TODO(hs): perform precheck to verify the retrieved AK certificate chain
+	// does belong to the TPM that's in use? Depending on how the certificate
+	// was obtained and stored, it might've been altered somehow.
+	// TODO(hs): support attestation flow with multiple Attestation CAs for a
+	// single AK? Currently an AK is identified just by its name and can only
+	// have one AK certificate (chain) signed by one Attestation CA at a time.
+
+	return ak, nil
+}
+
+// attestationStatement constructs and marshals the attestation
+// object for the `tpm` format.
+func attestationStatement(ctx context.Context, key *tpm.Key, akChain []*x509.Certificate) ([]byte, error) {
+	params, err := key.CertificationParameters(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed obtaining key certification parameters: %w", err)
+	}
+
+	akChainBytes := make([][]byte, len(akChain))
+	for i, cert := range akChain {
+		akChainBytes[i] = cert.Raw
+	}
+
+	obj := &attestationObject{
+		Format: "tpm",
+		AttStatement: map[string]interface{}{
+			"ver":      "2.0",
+			"alg":      int64(-257), // AlgRS256 (COSE identifier); depends on type of the private key
```

Comment:
We're also adding support for ECDSA keys, we should fix this.

Reply:
Good catch 🙂